src/Google.Cloud.Functions.Framework/GcfEvents/GcfConverters.cs

```diff
@@ -292,6 +294,63 @@ protected override void MaybeReshapeData(Request request)
                     request.Data["publishTime"] = timestamp.UtcDateTime.ToString(formatString, CultureInfo.InvariantCulture);
                 }
                 request.Data = new Dictionary<string, object> { { "message", request.Data } };
+
+                // The subscription is not provided in the legacy format, but *is* provided in
+                // the raw Pub/Sub push notification (including in the emulator) so we should use it if we've got it.
+                if (request.RawPubSubSubscription is string subscription)
+                {
+                    request.Data["subscription"] = subscription;
+                }
+            }
+
+            /// <summary>
+            /// Normalizes a raw Pub/Sub push notification (from either the emulator
+            /// or the real Pub/Sub service) into the existing legacy format.
+            /// </summary>
+            /// <param name="request">The incoming request body, parsed as a <see cref="Request"/> object</param>
+            /// <param name="path">The HTTP request path, from which the topic name will be extracted.</param>
+            internal static void NormalizeRawRequest(Request request, string path)
+            {
+                // Non-raw-Pub/Sub path: just a no-op
+                if (request.RawPubSubMessage is null && request.RawPubSubSubscription is null)
+                {
+                    return;
+                }
+                if (request.RawPubSubMessage is null || request.RawPubSubSubscription is null)
+                {
+                    throw new ConversionException("Request is malformed; it must contain both 'message' and 'subscription' properties or neither.");
+                }
+                if (request.Data is object ||
+                    request.Domain is object ||
+                    request.Context is object ||
+                    request.EventId is object ||
+                    request.EventType is object ||
+                    request.Params is object ||
+                    request.Timestamp is object)
+                {
+                    throw new ConversionException("Request is malformed; raw Pub/Sub request must object contain only 'message' and 'subscription' properties.");
```

Comment:
actually (sorry for not catching earlier):
```suggestion
                    throw new ConversionException("Request is malformed; raw Pub/Sub request must contain only 'message' and 'subscription' properties.");
```

Reply:
Whoops, thanks, fixed.